### PR TITLE
Button: always render the Tooltip component even when a tooltip should not be shown

### LINKED ISSUE #56522: `Button` creates new HTML elements every time it internally flips between showing tooltips or not
_As discussed originally in https://github.com/WordPress/gutenberg/discussions/44735_

### The problem

The `Button` component can, [based on which props are passed](https://github.com/WordPress/gutenberg/blob/aa7cc4d0404e0a830144c57e5211335b48f9ba2d/packages/components/src/button/index.tsx#L195-L206), show a tooltip when hovered or focused.

This feature is implemented by[ wrapping conditionally the button's contents in a `Tooltip` component](https://github.com/WordPress/gutenberg/blob/aa7cc4d0404e0a830144c57e5211335b48f9ba2d/packages/components/src/button/index.tsx#L259-L292).

The current implementation causes React not to be able to [reconciliate](https://react.dev/learn/preserving-and-resetting-state) the Button's contents if the need for showing a tootlip changes at runtime. In return, this causes the current `button` (or `a`) element rendered in the DOM to be removed and replaced by a new element, which causes focus loss.

### Known workarounds and potential solutions

The most clear way to avoid the re-render is to make sure that the `Tooltip` component is always rendered (instead of being rendered conditionally). This can be achieved in a few ways:

1. by specifying the `showTooltip` prop on `Button` (although this will force the tooltip to be always visible) ([example](https://github.com/WordPress/gutenberg/blob/ca2ee4d6fa8c7044cd1d3045c0985a77659f44fa/packages/editor/src/components/post-saved-state/index.js#L155-L161))
2. by avoiding to trigger the internal tooltip in `Button`, and instead wrapping `Button` externally with a `Tooltip`  (tested in #56502)
3. by tweaking the `Button` component to internally always render a `Tooltip` component (proposed in #56490)

Approaches 1 and 2 can work well to fix specific instances of `Button`, while approach 3 would solve the problem fundamentally for all `Button` usages (although it has the potential to introduce a breaking change).

A 4th approach (yet to be tested) would be to try to use the `key` special prop as a way to flag to React that a given component is stable (and therefore doesn't need to be destroyed and re-created across re-renders).

### Plan of action

- [x] Merge #56502
- [x] Work on a workaround for all `Button` components
    - [x] Try out the `key` method #56527
    - [x] Alternatively, understand if the solution proposed in #56490 is viable
- [x] Potentially undo previously added workarounds (like in #56502)

### PR DESCRIPTION
**Pull request by ciampo**
Fixes https://github.com/WordPress/gutenberg/issues/56522

Requires #57202, #57878, #57975, #58125

## What?

This PR tweaks the `Button` component so that it _always_ renders the `Tooltip` component, even when a tooltip shouldn't be shown.

It also un-does the temporary fix from https://github.com/WordPress/gutenberg/pull/56502

## Why?

This change is necessary to fix the issue described in https://github.com/WordPress/gutenberg/discussions/44735, where basically the element rendered by `Button` would be removed and re-added to the DOM any time the `Button` component switches from having to show a tooltip to not having to show it, and vice-versa.

The fact that a new HTML element is re-created every time triggers a focus loss that can be very annoying, especially to assistive technology users.

## How?

After some debugging, the issue seems related to the fact that React couldn't reconciliate the "with tooltip" and "without tooltip" version of the `Button` component without re-creating the whole button every time.

To solve this, I decided to tweak `Button` so that it always renders the `Tooltip` component.

_Note: this doesn't mean that the tooltip will always be shown, though — the original user-facing behaviour should have been retained by only passing props to the `Tooltip` component when needed._

### Testing Instructions for Keyboard

From https://github.com/WordPress/gutenberg/discussions/44735#discussioncomment-7639716: 

- Edit a draft post and make sure it's saved so that the Save button shows the text 'Saved'.
- Make any edit so that the button text changes to 'Save draft'.
- Select the Save button in the dev tools inspector.
- Right click and enable 'Break on > node removal'.
- Click the Save button.
- Observe the button doesn't gets removed from the DOM, and the keyboard focus stays on the button

A similar approach can also be used directly in Storybook.

Also, make sure that in general, tooltips show on `Button` when they are supposed to, and don't show when they are not supposed to.

### CODE REVIEW

File: packages/editor/src/components/post-saved-state/index.js
Comment (ciampo) [inline packages/editor/src/components/post-saved-state/index.js]: I will undo these changes before merging, but for now this helps with testing those changes

Comment (afercia) [reply packages/editor/src/components/post-saved-state/index.js]: Yes this is necessary for testing to reproduce the removal of the Button from the DOM, see https://github.com/WordPress/gutenberg/discussions/44735#discussioncomment-7639716

**Review by afercia — COMMENTED** [review]
Thanks @ciampo for working on this 🙏 
I can confirm that with this PR the focus loss doesn't occur any longer.

**Comment by ciampo** [comment]
It looks like some `Tooltip` unit tests are failing, I'll look into it.

**Comment by ciampo** [comment]
After taking a look, failures seem related to the fact that `Tooltip` tests use `Button` internally. Swapping the `Button` component with the `button` HTML element causes the tests to pass correctly.

With the changes from this PR, since `Button` always internally renders a `Tooltip`, rendering `<Tooltip><Button /></Tooltip>` effectively results in two `Tooltip` components being nested: `<Tooltip><Tooltip>{/* button contents */}</Tooltip></Tooltip>`.

I would argue that this scenario has been around on `trunk` for a while and that the changes from this PR have only made it easier to spot. At the same time, this PR would, unfortunately, introduce a regression to any consumer that is rendering `<Tooltip><Button /></Tooltip>`.

We have two main approaches that we could take:

- we can argue that `Button` was never meant to be rendered inside a `Tooltip`, since it already comes with Tooltip functionality. We tweak Tooltip unit tests to render an HTML `button`, and we merge this PR (knowing that it could introduce regressions for folks rendering `Button` inside a `Tooltip`.
- or we take a more conservative approach and avoid "amplifying" this regression (which is already on trunk, but harder to trigger). This means that we'll have to find an alternative approach to this fix. A short-term solution could be to render a `Tooltip` around the `Button` for the save button (to be tested).


What do folks think?

**Comment by ciampo** [comment]
> * A short-term solution could be to render a `Tooltip` around the `Button` for the save button (to be tested).

Alternative solution (specific to the save button) in https://github.com/WordPress/gutenberg/pull/56502

File: packages/components/src/button/index.tsx
Comment (tyxla) [inline packages/components/src/button/index.tsx]:
```suggestion
	// we don't pass the props to the `Tooltip` component.
```

File: packages/components/src/button/index.tsx
Comment (tyxla) [inline packages/components/src/button/index.tsx]: Do we really earn anything by memoizing these scalar props? It feels like the memorization here is unnecessary.

File: packages/editor/src/components/post-saved-state/index.js
Comment (tyxla) [reply packages/editor/src/components/post-saved-state/index.js]: Sounds good, let's make sure to not miss removing it before merging.

**Review by tyxla — COMMENTED** [review]
This is a fair compromise IMHO 👍 

> This PR would, unfortunately, introduce a regression to any consumer that is rendering `<Tooltip><Button /></Tooltip>`

Is it even possible that there is such an instance existing and it's not broken in the same way as the test breaks?

File: packages/components/src/button/index.tsx
Comment (ciampo) [reply packages/components/src/button/index.tsx]: Probably. I was testing different ways to avoid re-renders in the hop that it would help with React's reconciliation. I will try removing it

**Comment by ciampo** [comment]
> Is it even possible that there is such an instance existing and it's not broken in the same way as the test breaks?

Right now, a consumer may render `<Tooltip><Button /></Tooltip>` without issues, in case the `Button` doesn't also internally render a `Tooltip`. With the changes from this PR, `Button` would _always_ render a `Tooltip`, which causes the  `<Tooltip><Button /></Tooltip>` pattern to _never_ work.

---

> This is a fair compromise IMHO 👍

I will also try one more approach using [the `key` prop](https://react.dev/learn/rendering-lists) to force the hand with React's reconciliation, although I'm not sure if it will work.

**Comment by ciampo** [comment]
> I will also try one more approach using [the `key` prop](https://react.dev/learn/rendering-lists) to force the hand with React's reconciliation, although I'm not sure if it will work.

Opened https://github.com/WordPress/gutenberg/pull/56527, although I'm not managing to fix the issue.

We may have to resort to using the fix proposed in this PR. In order to mitigate the disruption, one potential fix may be to add some logic to the `Tooltip` component so that, if an instance of `Tooltip` detects being rendered inside another `Tooltip`, it will "disable" its tooltip functionality — so that only the topmost `Tooltip` actually renders a tooltip. We could also render a warning to the console letting devs know that a tooltip was suppressed because it was nested in another tooltip.

Let me know if this sounds like a good ide.

**Comment by tyxla** [comment]
> Right now, a consumer may render `<Tooltip><Button /></Tooltip>` without issues, in case the `Button` doesn't also internally render a `Tooltip`. With the changes from this PR, `Button` would _always_ render a `Tooltip`, which causes the `<Tooltip><Button /></Tooltip>` pattern to _never_ work.

This is definitely not great and could potentially break existing usages.

> We may have to resort to using the fix proposed in this PR. In order to mitigate the disruption, one potential fix may be to add some logic to the Tooltip component so that, if an instance of Tooltip detects being rendered inside another Tooltip, it will "disable" its tooltip functionality — so that only the topmost Tooltip actually renders a tooltip. We could also render a warning to the console letting devs know that a tooltip was suppressed because it was nested in another tooltip.

I think this sounds like a good compromise. Do we have any usages in the Gutenberg core that will be affected?

File: packages/components/src/button/index.tsx
Comment (ciampo) [inline packages/components/src/button/index.tsx:200]: Making sure that `shouldShowTooltip` is an actual `boolean`

File: packages/editor/src/components/post-saved-state/index.js
Comment (ciampo) [inline packages/editor/src/components/post-saved-state/index.js:1]: Changes in this file mostly undo the temporary fix from #56502

**Comment by ciampo** [comment]
Update:

- I opened [a PR](https://github.com/WordPress/gutenberg/pull/57202) with changes to support nested `Tooltip` components (ie. only the outer `Tooltip` component displays a tooltip)
- While testing changes, I noticed a regression related to the tooltip not hiding when the anchor is blurred. I opened [an issue](https://github.com/WordPress/gutenberg/issues/57206) and [a PR with a potential fix](https://github.com/WordPress/gutenberg/pull/57204)

Until #57202 (or any alternative solution) is merged, this PR will stay on hold.

**Comment by ciampo** [comment]
Rebased to include changes from https://github.com/WordPress/gutenberg/pull/57202. Will look into unit test failures — there may be some extra unintended consequences when nesting tooltips in `Button` (ie. class names being overwritten)

File: packages/components/src/tooltip/index.tsx
Comment (ciampo) [inline packages/components/src/tooltip/index.tsx]: @diegohaz I'd like to get a check from your side that this is a good approach — basically making sure that, when we clone `children` in inner nested tooltips (ie. the tooltip part gets ignored), we forward the correct props to the anchor.

I started by merging `className` and `styles`, then I realized that I needed to remove a few extra things added automatically by the context system (see the `removeExtraPropsAddedByContext` function). Finally, I had to chain events, and for now I chose to chain the ones that seem essential to keep the tooltip functioning correctly.

Please let me know if this approach makes sense to you, or if you have suggestions on how to improve it 🙏

File: packages/components/src/tooltip/index.tsx
Comment (ciampo) [inline packages/components/src/tooltip/index.tsx]: @diegohaz I'd like to get a check from your side that this is a good approach — basically making sure that, when we clone `children` in inner nested tooltips (ie. the tooltip part gets ignored), we forward the correct props to the anchor.

I started by merging `className` and `styles`, then I realized that I needed to remove a few extra things added automatically by the context system (see the `removeExtraPropsAddedByContext` function). Finally, I had to chain events, and for now I chose to chain the ones that seem essential to keep the `ariakit` tooltip functioning correctly.

Please let me know if this approach makes sense to you, or if you have suggestions on how to improve it 🙏

Comment (diegohaz) [reply packages/components/src/tooltip/index.tsx]: > @diegohaz I'd like to get a check from your side that this is a good approach — basically making sure that, when we clone `children` in inner nested tooltips (ie. the tooltip part gets ignored), we forward the correct props to the anchor.

I'm not certain I grasp the intent behind that. Could you share a code example where this is used and it caused a problem when all that merge logic wasn't applied?

> I chose to chain the ones that seem essential to keep the `ariakit` tooltip functioning correctly.

This is particularly concerning. Those events aren't part of Ariakit's public API. They could be taken out and new _essential_ event handlers might be introduced in patch versions.

Comment (ciampo) [reply packages/components/src/tooltip/index.tsx]: I recorded a video to give more context:


https://github.com/WordPress/gutenberg/assets/1083581/09bd2a3c-7cd9-43a9-bd66-d26b6a07416d

Comment (diegohaz) [reply packages/components/src/tooltip/index.tsx]: Thanks for the video; it was really helpful.

Could you try swapping `cloneElement` with `Ariakit.Role`? I did a quick run of this code in Storybook and it seems to be functioning as expected, though I might be overlooking something:

```jsx
		return isOnlyChild ? (
			<Ariakit.Role { ...clonedProps } render={ children } />
		) : (
			children
		);
```

This is essentially the same, but it automatically merges any function prop that starts with `on` (if conflicts arise). This should be a safer approach.

Comment (ciampo) [reply packages/components/src/tooltip/index.tsx]: Sounds good, thank you! I gave it a quick try and it seems to work as expected on my end too.

**Comment by ciampo** [comment]
> Rebased to include changes from #57202. Will look into unit test failures — there may be some extra unintended consequences when nesting tooltips in `Button` (ie. class names being overwritten)

There was indeed an issue with how props were forwarded / merged / passed to the `children` of the inner `Tooltip` components. The issue was investigated and discussed in [this conversation](https://github.com/WordPress/gutenberg/pull/56490#discussion_r1450512847), and I've now opened [a separate PR](https://github.com/WordPress/gutenberg/pull/57878) to test it and merge it in isolation.

I will rebase this PR on top of `trunk` once #57878 is merged

**Comment by ciampo** [comment]
With #57878, #57975 and #58125 merged, we should be finally ready for a final round of reviews.

I've also added [a unit test](https://github.com/WordPress/gutenberg/pull/56490/commits/03404d498e787ff7409ea93d92c3cbccaa3982a2) which should fail on `trunk` and pass on this PR.

File: packages/components/src/button/index.tsx
Comment (mirka) [inline packages/components/src/button/index.tsx:257]: Nice simplification here 👍

**Review by mirka — APPROVED** [review]
Code looks good, and is testing well 🚀

File: packages/components/src/button/test/index.tsx
Comment (tyxla) [inline packages/components/src/button/test/index.tsx:234]: Sorry for the silly question, but can an invisible element be focused? With that in mind, maybe `toHaveFocus()` is enough?

**Review by tyxla — APPROVED** [review]
Works well in my testing. Thanks @ciampo 🚀

File: packages/components/src/button/test/index.tsx
Comment (ciampo) [reply packages/components/src/button/test/index.tsx:234]: Not silly at all! I removed those extra assertions as they were redundant.

**Comment by ellatrix** [comment]
@ciampo Why would a button switch from having a tooltip to not having a tooltip while it's rendered in the DOM? What's the use case?

Edit: I read the testing instructions, so the save button :)

**Comment by ellatrix** [comment]
This is causing so much work to be done for buttons that never even show a tooltip. More often than not, buttons never have a tooltip. Why is this there even a tooltip option for a button, why is this not an optional wrapper component?

**Comment by mirka** [comment]
> More often than not, buttons never have a tooltip. Why is this there even a tooltip option for a button, why is this not an optional wrapper component?

FWIW, I would definitely not include a tooltip (and its floating-ui dependencies) if we were to design a Button component from scratch. Maybe in a v2!

**Comment by ciampo** [comment]
> This is causing so much work to be done for buttons that never even show a tooltip. More often than not, buttons never have a tooltip. Why is this there even a tooltip option for a button, why is this not an optional wrapper component?

@ellatrix These are all correct questions that we also asked ourselves. Unfortunately the decision to include the `Tooltip` component inside the `Button` component was made before I started working on Gutenberg; and the backwards compat policy forces us to keep around.

I agree with your observations and, as also Lena said, if we were to build a `Button` component from scratch,  we would keep the component very focused on button functionality —`Tooltip` functionality would be delegated to a separate `Tooltip` component.

**Pull request merged**
